src/tools/getDatasourceMetadata/datasourceMetadataUtils.ts

```diff
@@ -1,26 +1,68 @@
+import { z } from 'zod';
+
 import { GraphQLResponse } from '../../sdks/tableau/apis/metadataApi.js';
-import { MetadataResponse } from '../../sdks/tableau/apis/vizqlDataServiceApi.js';
+import {
+  FieldMetadata,
+  Function,
+  MetadataResponse,
+} from '../../sdks/tableau/apis/vizqlDataServiceApi.js';
+
+type Field = Partial<{
+  name: string;
+  dataType: FieldMetadata['dataType'];
+  defaultAggregation: z.infer<typeof Function>;
+  description: string;
+  descriptionInherited: Array<{
+    attribute: string;
+    value: string | null;
+  } | null>;
+  dataCategory: string;
+  role: string;
+  defaultFormat: string;
+  formula: string;
+  isAutoGenerated: boolean;
+  hasUserReference: boolean;
+  binSize: number;
+}>;
+
+type CombinedFieldsResult = {
+  fields: Array<Field>;
+};
 
-export function buildDataDictionary(
+export function combineFields(
   readMetadataResult: MetadataResponse,
   listFieldsResult: GraphQLResponse,
-): any {
-  // create a dataDictionary that combines field data from
+): CombinedFieldsResult {
+  // Create a response object that combines field data from
   // readMetadata (VizQL Data Service API) and listFields (GraphQL Metadata API) results
   // to optimize for LLM accuracy and reduce tokens in response.
-  const dataDictionary = {
-    fields: Array<any>(),
+  const combinedFields: CombinedFieldsResult = {
+    fields: [],
   };
 
   if (!readMetadataResult.data) {
-    // TODO: what should we do here?
-    return dataDictionary;
+    if (listFieldsResult.data.publishedDatasources[0]?.fields.length) {
+      // fallback to returning listFields results if we don't have any fields from readMetadata.
+      for (const field of listFieldsResult.data.publishedDatasources[0].fields) {
+        const toPush: Field = { name: field.name };
+        if (field.dataType?.length) {
+          toPush.dataType = field.dataType as Field['dataType'];
+        }
+        if (field.aggregation?.length) {
+          toPush.defaultAggregation = field.aggregation as FieldMetadata['defaultAggregation'];
+        }
+        populateFieldWithAdditionalProperties(field as Field, toPush);
```

Comment:
Consider removing the `as Field` type assertion and updating the `Field` type to allow for nulls.

```ts
type Field = Partial<{
  name: string;
  dataType: string | null;
  defaultAggregation: string | null;
  description: string | null;
  descriptionInherited: Array<{
    attribute: string;
    value: string | null;
  } | null> | null;
  dataCategory: string | null;
  role: string | null;
  defaultFormat: string | null;
  formula: string | null;
  isAutoGenerated: boolean | null;
  hasUserReference: boolean | null;
  binSize: number | null;
}>;
```